docs/adr/0006-consensus-governance.md

```diff
@@ -0,0 +1,459 @@
+  ```
+
+  And CBOR-serialized value:
+
+  ```golang
+  // Vote is a governance vote.
+  type Vote uint8
+
+  const (
+      VoteYes     Vote = 1
+      VoteNo      Vote = 2
+      VoteAbstain Vote = 3
+  )
+  ```
+
+- **List of pending upgrades (`0x84`)**
+
+  Each pending upgrade is stored under a separate storage key with the following
+  key format:
+
+  ```
+  0x84 <upgrade-epoch (beacon.EpochTime)> <proposal-id (uint64)>
+  ```
+
+### Genesis Document
+
+The genesis document needs to be updated to include a `governance` field with
+any initial state (see [_State_]) and consensus parameters (see [_Consensus
+Parameters_]) for the governance service.
+
+[_State_]: #state
+[_Consensus Parameters_]: #consensus-parameters
+
+### Transaction Methods
+
+This proposal adds the following transaction methods in the governance module:
+
+#### Submit Proposal
+
+Proposal submission enables a new consensus layer governance proposal to be
+created.
+
+**Method name:**
+
+```
+gov.SubmitProposal
+```
+
+**Body:**
+
+```golang
+// ProposalContent is a consensus layer governance proposal content.
+type ProposalContent struct {
+    Upgrade       *UpgradeProposal       `json:"upgrade,omitempty"`
+    CancelUpgrade *CancelUpgradeProposal `json:"cancel_upgrade,omitempty"`
+}
+
+// UpgradeProposal is an upgrade proposal.
+type UpgradeProposal struct {
+    upgrade.Descriptor
+}
+
+// CancelUpgradeProposal is an upgrade cancellation proposal.
+type CancelUpgradeProposal struct {
+    // ProposalID is the identifier of the pending upgrade proposal.
+    ProposalID uint64 `json:"proposal_id"`
+}
+```
+
+**Fields:**
+
+- `upgrade` (optional) specifies an upgrade proposal.
+- `cancel_upgrade` (optional) specifies an upgrade cancellation proposal.
+
+Exactly one of the proposal kind fields needs to be non-nil, otherwise the
+proposal is considered malformed.
+
+Upon processing any proposal the following steps are first performed:
+
+- The account indicated by the signer is loaded.
+
+- If the account balance is less than `min_proposal_deposit`, the method call
+  fails with `ErrInsufficientBalance`.
+
+Upon processing an **`UpgradeProposal`** the following steps are then performed:
+
+- The upgrade descriptor is checked for basic internal validity. If the check
+  fails, the method call fails with `ErrInvalidArgument`.
+
+- The upgrade descriptor's `epoch` field is compared with the current epoch. If
+  the specified epoch is not at least `upgrade_min_epochs_diff` epochs ahead of
+  the current epoch, the method call fails with `ErrUpgradeTooSoon`.
+
+- The set of pending upgrades is checked to make sure that no upgrades are
+  currently pending. If there is an existing upgrade pending, the method call
+  fails with `ErrUpgradeAlreadyPending`.
+
+Upon processing a **`CancelUpgradeProposal`** the following steps are then
+performed:
+
+- The set of pending upgrades is checked to make sure that the given upgrade
```

Comment:
Who can cancel a specific proposal, I assume only the one initially submitting the proposal right?

There is currently nothing preventing last-second canceling of a proposal? Should there be something like that? (e.g. `cancel_min_epochs_diff`). Alternatively should the `CancelUpgradeProposal` require voting as well?

Additionally, I think I'm missing some context on why even having the cancel proposal, what's a common scenario where it would be needed? (things that come to mind are late-discovered bugs in the proposed upgrade?)

Reply:
> Who can cancel a specific proposal, I assume only the one initially submitting the proposal right?

Anyone (that can fund a deposit) can propose an upgrade cancellation. It needs to go through the usual governance process.

> Additionally, I think I'm missing some context on why even having the cancel proposal, what's a common scenario where it would be needed? Things that come to mind are late-discovered bugs in the proposed upgrade? anything else?

Yes, that is exactly the use case for having an upgrade cancellation proposal. I don't think there are any other scenarios.

Reply:
> Alternatively should the CancelUpgradeProposal require voting as well?

Any proposal, including the cancel upgrade proposal, requires the full governance process, so voting.

> There is currently nothing preventing last-second canceling of a proposal? Should there be something like that? (e.g. cancel_min_epochs_diff).

Yeah that's probably a good idea, I'll update the ADR.

Reply:
Makes sense, thanks. Yeah disregard the comment about requiring voting, i see that this results in a proposal that goes through the governance process :+1: 

So I think there should probably be an additional check here that the resulting proposal end height (`current_height` + `voting_period`) is not after the start of the epoch in which the `Upgrade` is taking place?

Reply:
> So I think there should probably be an additional check here that the resulting proposal end height (current_height + voting_period) is not after the start of the epoch in which the Upgrade is taking place?

Either that or a sanity check that `voting_period` must be less than any of the `*_min_epochs_diff`?

Reply:
> Although with variable epoch lengtths that are coming i don't think this is possible?

Oh good point, so we could also change the voting period to be in epochs instead?

Reply:
Voting period in epochs makes sense to me.

Reply:
Done.